### PR TITLE
Tor & Wenjie - VideoStore - Octos 

### PR DESCRIPTION
**Pull request by torshimizu**
# Video Store API
Congratulations! You're submitting your assignment!
If you didn't get to the functionality the question is asking about, reply with what you would have done if you had completed it.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Explain how you came up with the design of your ERD, based on the seed data. | At first we did not use the seed data, so we had to go back and add some fields to our models (like movies checked out count). We worked off of the project README, and built our relations that way.
Describe a set of positive and negative test cases you implemented for a model. |  Positive test case: A movie can be created with sufficient data. Negative test case: a movie cannot be created without a title or with a negative inventory. 
Describe a set of positive and negative test cases you implemented for a controller. | Positive test case: For show, a single movie's details will be shown when the movie is found. Negative test case: a not found status would be sent with error text if the movie was not found.
How does your API respond when bad data is sent to it? | If bad data is sent, we respond with error text in a hash and send a status code of bad_request or not_found (if the id was not found).
Describe one of your custom model methods and why you chose to wrap that functionality into a method. | We have a method to find a rental with a given movie and customer id that has not yet been checked in. This method is used by our Rentals controller for when a movie is checked in. We decided to put this in the model so that we can do further testing and it can also be reused if we ever needed to modify or update the rental information.
Do you have any recommendations on how we could improve this project for the next cohort? | Do something different than a Video Store - something more fun.
Link to Trello | https://trello.com/b/yLm7THn3/video-store-api
Link to ERD | https://www.lucidchart.com/documents/edit/6cce0be7-55f5-4c57-9b0e-6d6c83d5c2e6

### CODE REVIEW

File: app/models/customer.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  validates :name, presence: true
+  validates :phone, presence: true
+  validate :same_name_and_same_phone, on: :create
+
+  has_many :rentals
+
+  def same_name_and_same_phone
+    customers = Customer.where(name: name)
+
+    if customers.any? {|customer| customer.phone == phone}
```
Comment (droberts-sea) [inline app/models/customer.rb:11]: Rails supports this natively:

```
validates :phone, uniqueness: { scope: :name }
```

File: app/models/rental.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  validate :has_enough_inventory?, if: :movie_id?
+
+  def movie_id?
+    return self.movie_id
+  end
+
+  def has_enough_inventory?
+    unless self.movie.available_inventory > 0
+      errors[:quantity] << 'Not enough inventory for this rental'
+    end
```
Comment (droberts-sea) [inline app/models/rental.rb:18]: Nice work on this custom validation

File: app/models/rental.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def self.find_checked_out_movie(movie_id, customer_id)
+    rental = Rental.where(movie_id: movie_id, customer_id: customer_id, check_in_date: nil)
+    return rental.empty? ? nil : rental.first
+  end
```
Comment (droberts-sea) [inline app/models/rental.rb:24]: Good use of model methods here. I might take it a step further and have something like `Rental.checkout(movie_id, customer_id)` and `Rental.checkin(movie_id, customer_id)`, which encapsulate all of the work in one place. You could even wrap the whole thing in a transaction.

File: app/models/rental.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+
+  def self.update_movie_and_customer(rental)
+
+    if rental.check_in_date.nil?
+      new_inventory = rental.movie.available_inventory - 1
```
Comment (droberts-sea) [inline app/models/rental.rb:30]: This feels like it should be an instance method, since it requires an instance to run.

File: test/controllers/rentals_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      movies(:one).available_inventory.must_equal movie_inventory - 1
+      body = JSON.parse(response.body)
+      body.must_be_kind_of Hash
+      Rental.find(body['id']).customer_id.must_equal rental_data[:customer_id]
+      Rental.count.must_equal @old_rental_count + 1
+    end
+
+    it 'returns bad request and error text for bad data' do
+      rental_data = {
+        movie_id: movies(:one).id,
```
Comment (droberts-sea) [inline test/controllers/rentals_controller_test.rb:29]: What if the movie doesn't have enough inventory left in stock?

File: test/controllers/rentals_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+      body = JSON.parse(response.body)
+      body.must_be_kind_of Hash
+      Rental.find(body['id']).check_in_date.must_equal Date.today
+    end
+
+    it "return bad request and error if the rental DNE" do
+      rental = rentals(:one)
+      rental_data = {
+        movie_id: rental.movie_id + 100,
```
Comment (droberts-sea) [inline test/controllers/rentals_controller_test.rb:76]: What about a rental that has already been checked in? What if there are multiple rentals for the same movie-customer pair?

File: test/models/rental_test.rb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        customer_id = rental.customer_id
+        result = Rental.find_checked_out_movie(movie_id, customer_id)
+        result.must_be_kind_of Rental
+        result.must_equal rental
+      end
+
+      it "returns nil if the rental DNE" do
+        rental = rentals(:one)
+        movie_id = rental.movie_id + 100
+        customer_id = rental.customer_id
```
Comment (droberts-sea) [inline test/models/rental_test.rb:62]: What if the rental has been checked in?

**Comment by droberts-sea** [comment]
# Video Store
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene | yes
Comprehension questions | yes
**General** |
Business Logic in Models | yes - good work
All required endpoints return expected JSON data | yes
Requests respond with appropriate HTTP Status Code | yes
Errors are reported | yes
**Testing** |
Passes all Smoke Tests | yes
Model Tests - all relations, validations, and custom functions test positive & negative cases | yes
Controller Tests - URI parameters and data in the request body have positive & negative cases | some edge cases missing, but generally good.
**Overall** | Good work overall.
